llvm/lib/Transforms/Utils/CodeLayout.cpp

```diff
@@ -734,14 +743,25 @@ class ExtTSPImpl {
         // Get candidates for merging with the current chain.
         for (const auto &[ChainSucc, Edge] : ChainPred->Edges) {
           // Ignore loop edges.
-          if (ChainPred == ChainSucc)
+          if (Edge->isSelfEdge())
             continue;
-
           // Stop early if the combined chain violates the maximum allowed size.
           if (ChainPred->numBlocks() + ChainSucc->numBlocks() >= MaxChainSize)
             continue;
+          // Don't merge the chains if they have vastly different densities.
+          // We stop early if the ratio between the densities exceeds
+          // MaxMergeDensityRatio. Smaller values of the option result in
+          // fewer merges (hence, more chains), which in turn typically yields
+          // smaller size of the hot code section.
```

Comment:
I am confused by this comment. How can the hot section get smaller?  Yes, we don't merge the chains, but they still get concatenated at the end.

Reply:
I adjusted the comment (and the one above). Please check if the new one makes sense. Thanks